### PR TITLE
card number/ bin validity check

### PR DESCRIPTION
**Pull request by jaimejiyepark**
note:
* changed all argument names to `cardNumber` to stay consistent

### CODE REVIEW

**Comment by jaimejiyepark** [comment]
I changed `luhnCheck` and `isValidBin` to non-public since they are functions used in conjunction with each other in the greater `isValidNumber` function. I noticed that `luhnCheck` was being used in `RecognizeNumbers.swift` with checking the `number.count` which is what `isValidNumber` does. Also, `isValidBin` in the context of cards while in `CardVerify` we use it to get the card icon images. To remedy that I made a function `determineCardNetwork` public for use. I also followed Adam's android code pretty closely.

File: CardScan/Classes/CreditCardUtils.swift [hunk 1]
```diff
@@ -1,29 +1,43 @@
 import Foundation
 
+public enum CardNetwork: String {
```
Comment (kingst) [inline CardScan/Classes/CreditCardUtils.swift]: does this need to be public?

File: CardScan/Classes/CreditCardUtils.swift [hunk 2]
```diff
@@ -1,29 +1,43 @@
...
+    }
     
     // https://en.wikipedia.org/wiki/Luhn_algorithm
     // assume 16 digits are for MC and Visa (start with 4, 5) and 15 is for Amex
     // which starts with 3
-    public static func luhnCheck(_ cardNumber: String) -> Bool {
-        if cardNumber.count == 0 {
-            return false
-        } else if !isValidBin(number: cardNumber) {
+    static func isValidLuhnNumber(cardNumber: String) -> Bool {
```
Comment (kingst) [inline CardScan/Classes/CreditCardUtils.swift]: You're taking a function that used to be public and making it not public. If people who use our library use this function they're going to be upset that they need to change code now.

File: CardScan/Classes/CreditCardUtils.swift [hunk 3]
```diff
@@ -40,68 +54,76 @@ public struct CreditCardUtils {
         return sum % 10 == 0
     }
     
-    public static func isValidBin(number: String) -> Bool {
```
Comment (kingst) [inline CardScan/Classes/CreditCardUtils.swift]: Same here

File: CardScan/Classes/CreditCardUtils.swift [hunk 4]
```diff
@@ -40,68 +54,76 @@ public struct CreditCardUtils {
...
     }
     
-    public static func isAmex(number: String) -> Bool {
-        let prefix = String(number.prefix(2))
-        return number.count == maxPanLengthAmericanExpress && prefixesAmericanExpress.contains(prefix)
+    static func isValidLength(cardNumber: String) -> Bool {
+        return isValidLength(cardNumber: cardNumber, network: determineCardNetwork(cardNumber: cardNumber))
     }
-    
-    public static func isUnionPay(number: String) -> Bool {
```
Comment (kingst) [inline CardScan/Classes/CreditCardUtils.swift]: Same with all of the isSomeBank functions

File: CardScan/Classes/CreditCardUtils.swift [hunk 5]
```diff
@@ -40,68 +54,76 @@ public struct CreditCardUtils {
...
+            return cardNumberLength == maxPanLengthDinersClub
+        default:
+            return cardNumberLength == maxPanLength
+        }
     }
     
-    public static func isJcb(number: String) -> Bool {
-        let prefix = String(number.prefix(2))
-        return number.count == maxPanLength && prefixesJcb.contains(prefix)
+    public static func determineCardNetwork(cardNumber: String) -> CardNetwork {
```
Comment (kingst) [inline CardScan/Classes/CreditCardUtils.swift]: Ok, I see, ignore my comment above

File: CardScan/Classes/CreditCardUtils.swift [hunk 6]
```diff
@@ -40,68 +54,76 @@ public struct CreditCardUtils {
...
+    static func hasAnyPrefix( cardNumber: String, prefixes: [String] ) -> Bool {
+        return prefixes.filter { cardNumber.hasPrefix($0) }.count > 0
     }
     
-    public static func format(number: String) -> String {
-        if number.count == 16 {
-            return format16(number: number)
-        } else if number.count == 15 {
-            return format15(number: number)
+    public static func format(cardNumber: String) -> String {
```
Comment (kingst) [inline CardScan/Classes/CreditCardUtils.swift]: if people are using this function they have to change their source code after this change, which we should avoid for changes like this

**Comment by jaimejiyepark** [comment]
No it doesn't change anything written in the documentation. It only changes the internal interactions.

**Comment by jaimejiyepark** [comment]
okay real quick though, can i add the `isValidCVV` and `isValidExpirationDate` logic to this too. It's originally in `CardVerify`. Doesn't affect the `CardVerify` docs

File: CardScan/Classes/CreditCardUtils.swift (hunk 1 above)
Comment (jaimejiyepark) [reply CardScan/Classes/CreditCardUtils.swift]: its public so it can be used in `CardVerify`

**Comment by jaimejiyepark** [comment]
Ill just make that a separate PR

**Pull request merged**